### PR TITLE
add correlation ids

### PR DESCRIPTION
**Pull request by PrathyushaLakkireddy**
Fixes https://github.com/allinbits/demeris-backend/issues/346

### CODE REVIEW

**Comment by sgerogia** [comment]
If you have not seen it, @gsora has left a relevant comment [in the issue](https://github.com/allinbits/demeris-backend/issues/346#issuecomment-1003944214).

Please replace the existing generated error id to be the `int_correlation_id`

**Comment by 0xbala-k** [comment]
> If you have not seen it, @gsora has left a relevant comment [in the issue](https://github.com/allinbits/demeris-backend/issues/346#issuecomment-1003944214).
> 
> Please replace the existing generated error id to be the `int_correlation_id`

Yup! Added that in the `Deps.WriteError` function [here](https://github.com/allinbits/demeris-api-server/blob/1b451c7fdea6e888bd5ccb62b518caf2872b3640/api/router/deps/deps.go#L51).

File: api/router/correlation.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+func CorrelationIDMiddleware(l *zap.SugaredLogger) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		addCorrelationID(c, l)
+	}
+}
+
+func addCorrelationID(c *gin.Context, l *zap.SugaredLogger) {
+	ctx := c.Request.Context()
+
+	correlationID := c.Request.Header.Get("X-Correlation-id")
```
Comment (sgerogia) [inline api/router/correlation.go]: Even though HTTP headers are case-insesitive, it makes sense to use a const for the header name

**Comment by sgerogia** [comment]
@SpideyPool192 I do not think that [this merge](https://github.com/allinbits/emeris-utils/pull/25) automatically created a release
Can you create a minor release of `emeris-utils` and update the code here?

This needs to be reviewed and merged today if possible

**Review by sgerogia — CHANGES_REQUESTED** [review]

File: api/router/correlation_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/cockroachdb/cockroach-go/v2/testserver"
+	"github.com/gofrs/uuid"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest/observer"
+)
+
+type chainsResponse struct {
+	Chains []supportedChain `json:"chains"`
+}
```
Comment (DeshErBojhaa) [inline api/router/correlation_test.go]: Hi, where it's used?

File: api/router/correlation_test.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+type supportedChain struct {
+	ChainName   string `json:"chain_name"`
+	DisplayName string `json:"display_name"`
+	Logo        string `json:"logo"`
+}
+
+func TestCorrelationIDMiddleWare(t *testing.T) {
```
Comment (DeshErBojhaa) [inline api/router/correlation_test.go:21]: Consider t.Parallel()?

File: api/router/correlation_test.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	go r.Serve(cfg.ListenAddr)
+	time.Sleep(2 * time.Second)
+
+	client := http.Client{
+		Timeout: 2 * time.Second,
+	}
+	req, err := http.NewRequest("GET", fmt.Sprintf("http://%s%s", cfg.ListenAddr, "/chains"), nil)
+	require.NoError(t, err)
+
+	id, _ := uuid.NewV4()
```
Comment (DeshErBojhaa) [inline api/router/correlation_test.go]: Why ignoring the err?

**Review by DeshErBojhaa — CHANGES_REQUESTED** [review]
Hi, can you write in short why you're panicking in the router instead of writing the error to log (even to the response maybe).

Also please shortly explain the logic behind the panic in deps.go

File: api/router/correlation_test.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/allinbits/emeris-utils/logging"
+	"github.com/allinbits/emeris-utils/store"
+	"github.com/cockroachdb/cockroach-go/v2/testserver"
+	"github.com/gofrs/uuid"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest/observer"
+)
+
+type chainsResponse struct {
```
Comment (sahith-narahari) [inline api/router/correlation_test.go]: unused var

File: api/router/correlation_test.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	Chains []supportedChain `json:"chains"`
+}
+
+type supportedChain struct {
```
Comment (sahith-narahari) [inline api/router/correlation_test.go]: I don't see usages for this one either, are we using these structs somewhere?

**Review by sahith-narahari — CHANGES_REQUESTED** [review]

File: api/router/router.go [hunk 7]
```diff
@@ -89,6 +91,14 @@ func (r *Router) Serve(address string) error {
 	return r.g.Run(address)
 }
 
+func (r *Router) setLoggerFromContext(c *gin.Context) {
+	l, err := logging.GetLoggerFromContext(c)
+	if err != nil && r.l == nil {
+		panic("cant get logger from context")
+	}
+	r.l = l
```
Comment (pitasi) [inline api/router/router.go]: is this safe to do? When multiple API calls arrive aren't them executed concurrently and sharing the same router object?

even it this was safe looks like an anti-pattern to me: request-related stuff should stay inside request's context. What's this r.l used for anyway?

**Review by pitasi — CHANGES_REQUESTED** [review]

File: api/router/correlation_test.go (hunk 2 above)
Comment (0xbala-k) [reply api/router/correlation_test.go]: Used this earlier. Forgot to remove.

File: api/router/correlation_test.go (hunk 6 above)
Comment (0xbala-k) [reply api/router/correlation_test.go]: Used this earlier. Forgot to remove.

**Comment by 0xbala-k** [comment]
> Hi, can you write in short why you're panicking in the router instead of writing the error to log (even to the response maybe).
> 
> Also please shortly explain the logic behind the panic in deps.go



> Hi, can you write in short why you're panicking in the router instead of writing the error to log (even to the response maybe).
> 
> Also please shortly explain the logic behind the panic in deps.go

The issue suggests that we use a logger from the context if it doesn't exist check if there is another logger (to the router) provided. If there is no logger found we panic.

In deps.go we are trying to replace the error id with the internal correlation id which we created earlier. Hence, we try to find it in the context and panic if not found.

File: api/router/router.go (hunk 7 above)
Comment (0xbala-k) [reply api/router/router.go]: We set a modified logger to the context in the `CorrelationIDMiddleware` function. Intention is to use this logger everywhere. As the api server uses the logger from router, I'm changing the router logger (r.l) to use the logger from context.

**Review by DeshErBojhaa — APPROVED** [review]
LGTM!

**Review by gsora — APPROVED** [review]

File: api/router/router.go (hunk 7 above)
Comment (pitasi) [reply api/router/router.go]: > We set a modified logger to the context in the CorrelationIDMiddleware function. Intention is to use this logger everywhere.

that's ok

> As the api server uses the logger from router, I'm changing the router logger (r.l) to use the logger from context.

I don't get why we need a router logger (+ I'm still concerned about concurrency safety)

**Review by akhilkumarpilli — APPROVED** [review]

File: api/router/router.go (hunk 7 above)
Comment (0xbala-k) [reply api/router/router.go]: cc @gsora @sahith-narahari

**Pull request merged**